Fix or disable randomly failing test in `RuntimeEventListenerSpec`

Testing the log messages is super unreliable and randomly fails in CI.

Try to find out why, and if it is not fixable, disable the test (`ignore`)

```
- must return valid log messages for each status event *** FAILED *** (111 milliseconds)
  java.lang.AssertionError: Received 1 excess messages on LoggingTestKitImpl(1,Some(INFO),None,None,Some(Initializing Agents and Services for Simulation ... ),None,None,Map(),true,None).
  at org.apache.pekko.actor.testkit.typed.internal.LoggingTestKitImpl.expect(LoggingTestKitImpl.scala:115)
  at edu.ie3.simona.event.listener.RuntimeEventListenerSpec.$anonfun$new$5(RuntimeEventListenerSpec.scala:154)
  at edu.ie3.simona.event.listener.RuntimeEventListenerSpec.$anonfun$new$5$adapted(RuntimeEventListenerSpec.scala:150)
  at scala.collection.immutable.List.foreach(List.scala:333)
  at edu.ie3.simona.event.listener.RuntimeEventListenerSpec.$anonfun$new$4(RuntimeEventListenerSpec.scala:150)
  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
  at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  ...
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -44,6 +44,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Logging wrong duration in the first simulation hour [#705](https://github.com/ie3-institute/simona/issues/705)
 - Fixed some compiler warnings [#657](https://github.com/ie3-institute/simona/issues/657)
 - Fixing false negative in ref system voltage validation [#706](https://github.com/ie3-institute/simona/issues/706)
+- Fixing randomly failing test in `RuntimeEventListenerSpec` etc. [#709](https://github.com/ie3-institute/simona/issues/709)
 
 ## [3.0.0] - 2023-08-07
 

--- a/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerKafkaSpec.scala
+++ b/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerKafkaSpec.scala
@@ -128,13 +128,14 @@ class RuntimeEventListenerKafkaSpec
       forAll(cases) { case (event, expectedMsg) =>
         listenerRef ! event
 
-        val records: List[SimonaEndMessage] =
-          eventually(timeout(20 seconds), interval(1 second)) {
+        eventually(timeout(20 seconds), interval(1 second)) {
+          val records =
             testConsumer.poll((1 second) toJava).asScala.map(_.value()).toList
-          }
 
-        records should have length 1
-        records should contain(expectedMsg)
+          records should have length 1
+          records should contain(expectedMsg)
+        }
+
       }
 
     }

--- a/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerKafkaSpec.scala
+++ b/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerKafkaSpec.scala
@@ -128,13 +128,20 @@ class RuntimeEventListenerKafkaSpec
       forAll(cases) { case (event, expectedMsg) =>
         listenerRef ! event
 
-        eventually(timeout(20 seconds), interval(1 second)) {
-          val records =
-            testConsumer.poll((1 second) toJava).asScala.map(_.value()).toList
+        val receivedRecord =
+          eventually(timeout(20 seconds), interval(1 second)) {
+            val records =
+              testConsumer.poll((1 second) toJava).asScala.map(_.value()).toList
 
-          records should have length 1
-          records should contain(expectedMsg)
-        }
+            // run until one record is received. After each second, if no record
+            // was received, the length check below fails and we retry
+            records should have length 1
+
+            // return final record to be checked
+            records.headOption.value
+          }
+
+        receivedRecord shouldBe expectedMsg
 
       }
 

--- a/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerSpec.scala
+++ b/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerSpec.scala
@@ -147,8 +147,12 @@ class RuntimeEventListenerSpec
         ),
       )
 
+      val loggingTestKit = LoggingTestKit.empty
+        // logger name for ActorContext loggers (ctx.log) is the class name
+        .withLoggerName(RuntimeEventListener.getClass.getName)
+
       events.foreach { case (event, level, msg) =>
-        LoggingTestKit.empty
+        loggingTestKit
           .withLogLevel(level)
           .withMessageContains(msg)
           .expect {

--- a/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerSpec.scala
+++ b/src/test/scala/edu/ie3/simona/event/listener/RuntimeEventListenerSpec.scala
@@ -47,30 +47,27 @@ class RuntimeEventListenerSpec
     with PrivateMethodTester {
 
   // global variables
-  val eventQueue = new LinkedBlockingQueue[RuntimeEvent]()
   val startDateTimeString = "2011-01-01 00:00:00"
   val endTick = 3600
   val duration = 10805000
-  val errMsg =
-    "Und wenn du lange in einen Abgrund blickst, blickt der Abgrund auch in dich hinein"
+  val errMsg = "testing error msg"
 
-  // to change for Ready event test cases
   val currentTick = 0
-
-  // build the listener
-  private val listenerRef = spawn(
-    RuntimeEventListener(
-      SimonaConfig.Simona.Runtime.Listener(
-        None,
-        None,
-      ),
-      Some(eventQueue),
-      startDateTimeString,
-    )
-  )
 
   "A runtime event listener" must {
     "add a valid runtime event to the blocking queue for further processing" in {
+      val eventQueue = new LinkedBlockingQueue[RuntimeEvent]()
+
+      val listenerRef = spawn(
+        RuntimeEventListener(
+          SimonaConfig.Simona.Runtime.Listener(
+            None,
+            None,
+          ),
+          Some(eventQueue),
+          startDateTimeString,
+        )
+      )
 
       //  valid runtime events
       val eventsToQueue: Seq[RuntimeEvent] = List(
@@ -96,6 +93,17 @@ class RuntimeEventListenerSpec
     }
 
     "return valid log messages for each status event" in {
+
+      val listenerRef = spawn(
+        RuntimeEventListener(
+          SimonaConfig.Simona.Runtime.Listener(
+            None,
+            None,
+          ),
+          None,
+          startDateTimeString,
+        )
+      )
 
       def calcTime(curTick: Long): String = {
         TimeUtil.withDefaults.toString(

--- a/src/test/scala/edu/ie3/simona/integration/RunSimonaStandaloneIT.scala
+++ b/src/test/scala/edu/ie3/simona/integration/RunSimonaStandaloneIT.scala
@@ -126,28 +126,25 @@ class RunSimonaStandaloneIT
   private def checkRuntimeEvents(
       runtimeEvents: Iterable[RuntimeEvent]
   ): Unit = {
-    runtimeEvents.toVector.size shouldBe 11
-    val groupedRuntimeEvents = runtimeEvents.toVector.groupBy {
-      case Initializing            => Initializing
-      case InitComplete(_)         => InitComplete
-      case Simulating(_, _)        => Simulating
-      case CheckWindowPassed(_, _) => CheckWindowPassed
-      case Done(_, _, _)           => Done
-      case other                   => fail(s"Unexpected runtime event: $other")
-    }
+    val groupedRuntimeEvents = runtimeEvents.groupBy(event => event.getClass)
 
-    groupedRuntimeEvents.size shouldBe 5
-    groupedRuntimeEvents.keySet should contain allOf (Simulating, CheckWindowPassed, InitComplete, Initializing, Done)
+    groupedRuntimeEvents.keySet should contain only (
+      classOf[Simulating],
+      classOf[CheckWindowPassed],
+      classOf[InitComplete],
+      classOf[Initializing.type],
+      classOf[Done]
+    )
 
     groupedRuntimeEvents
-      .get(Simulating)
+      .get(classOf[Simulating])
       .foreach(simulatingEvents => {
         simulatingEvents.size shouldBe 1
         simulatingEvents.headOption.foreach(_ shouldBe Simulating(0, 7200))
       })
 
     groupedRuntimeEvents
-      .get(CheckWindowPassed)
+      .get(classOf[CheckWindowPassed])
       .foreach(checkWindowsPassed => {
         checkWindowsPassed.size shouldBe 7
         checkWindowsPassed.foreach {
@@ -161,19 +158,19 @@ class RunSimonaStandaloneIT
       })
 
     groupedRuntimeEvents
-      .get(InitComplete)
+      .get(classOf[InitComplete])
       .foreach(initComplets => {
         initComplets.size shouldBe 1
       })
 
     groupedRuntimeEvents
-      .get(Initializing)
+      .get(classOf[Initializing.type])
       .foreach(initializings => {
         initializings.size shouldBe 1
       })
 
     groupedRuntimeEvents
-      .get(Done)
+      .get(classOf[Done])
       .foreach(dones => {
         dones.size shouldBe 1
         dones.headOption.foreach {

--- a/src/test/scala/edu/ie3/simona/model/participant/load/LoadModelScalingSpec.scala
+++ b/src/test/scala/edu/ie3/simona/model/participant/load/LoadModelScalingSpec.scala
@@ -340,7 +340,7 @@ class LoadModelScalingSpec extends UnitSpec with TableDrivenPropertyChecks {
         getRelativeDifference(
           quantile95,
           targetMaximumPower,
-        ) should be < Percent(1d)
+        ) should be < Percent(2d)
       }
 
       "correctly account for the scaling factor when targeting at maximum power" in {

